subprojects/execution/src/main/java/org/gradle/internal/execution/steps/IdentifyStep.java

```diff
@@ -51,42 +56,75 @@ public <T> Deferrable<Try<T>> executeDeferred(UnitOfWork work, C context, Cache<
 
     @Nonnull
     private IdentityContext createIdentityContext(UnitOfWork work, C context) {
-        InputFingerprinter.Result inputs = work.getInputFingerprinter().fingerprintInputProperties(
-            ImmutableSortedMap.of(),
-            ImmutableSortedMap.of(),
-            ImmutableSortedMap.of(),
-            ImmutableSortedMap.of(),
-            work::visitIdentityInputs
+        String workClassName = work.getClass().getName();
+        return operation(operationContext -> {
+                InputFingerprinter.Result inputs = work.getInputFingerprinter().fingerprintInputProperties(
+                    ImmutableSortedMap.of(),
+                    ImmutableSortedMap.of(),
+                    ImmutableSortedMap.of(),
+                    ImmutableSortedMap.of(),
+                    work::visitIdentityInputs
+                );
+                ImmutableSortedMap<String, ValueSnapshot> identityInputProperties = inputs.getValueSnapshots();
+                ImmutableSortedMap<String, CurrentFileCollectionFingerprint> identityInputFileProperties = inputs.getFileFingerprints();
+
+                Identity identity = work.identify(identityInputProperties, identityInputFileProperties);
+                String identityUniqueId = identity.getUniqueId();
+                operationContext.setResult(new Operation.Result() {
+                    @Override
+                    public String getIdentityUniqueId() {
+                        return identityUniqueId;
```

Comment:
I'd also just call this `getIdentity()`.

Reply:
I named it like this, because the `Identity` itself is not just a string, but an interface, which I suppose is future-proofing. So, I tried to make the details value to correspond to the particular member of the `Identity`. Given your suggestion, I suppose renaming it to be simply `getIdentity()` should work too.

Reply:
`Identity` is internal to the execution engine, as we are not exposing it via the build operations. In the build operation context there is only one identity, and it's this `String`. So I think we should call it just `identity`.

Reply:
Nit: you could also inline `dentityUniqueId` since `IdentityContext` is immutable. That way you don't need to come up with a name for the unique ID.

Reply:
I would prefer not to do this, because doing that extends the lifetime of the `IdentityContext` for as long as the operation result is alive. It is probably a non-substantial memory leak, but I would prefer to avoid unnecessary dependencies like these.